Allow certificate with duplicate principals in truststore.

divinehawk opened the pull request:
The previous implementation uses principal as a key for a hashmap. This change uses the certificate alias as the key. Alias is guaranteed to be unique vice the principal name which is only unique in combination with the public key.

Closes #33125

mposolda commented:
@divinehawk Thanks for the PR! The idea of the change looks good to me, but there are failing tests, which looks like a regression of this PR ( `X509OCSPResponderFailOpenTest` and `X509OCSPResponderTest`). This will need to be fixed.

divinehawk commented:
With these changes, the map is now  Map<X500Principal, List<X509Certificate>>, in order to preserve the other validations (CRL, OSCP, etc.). This required some additional logic in those modules in order to handle a list of certificates rather than just one. 

However, the matching of an issuer to a certificate is now improved since there is now signature verification rather than just the principal name.

Comments appreciated.

divinehawk committed:
Allow certificate with duplicate principals in truststore.

The previous implementation uses principal as a key for a hashmap storing one certificate per entry. To preserve lookups, the value is now a List of certificates.

Additional logic was added to build certification validation chains using signature verification rather than just principal.

Closes #33125

Signed-off-by: Matt Eaton <git@divinehawk.com>

mposolda reviewed (approved):
@divinehawk Approving, Thanks! Hopefully tests are OK after re-run
